Container provider multi endpoint

yaacov opened the pull request:
Convert the container provider new/edit forms to play nice with manageIQ new forms format (angular), and add multi endpoints (default and hawkular).

1. Convert provider form to angular
2. Add hawkular endpoint
3. Use new/update routes like in ems_cloud
4. Use hawkular endpoint to collect metrics

**new**
![manageiq containers providers-1](https://cloud.githubusercontent.com/assets/2181522/15145538/da497bc6-16be-11e6-9759-243d0bdc572e.png)

**default endpoint**
![manageiq containers providers-2](https://cloud.githubusercontent.com/assets/2181522/15145540/de4d486a-16be-11e6-84f6-fb68392c2953.png)

**hawkular endpoint**
![manageiq containers providers](https://cloud.githubusercontent.com/assets/2181522/15315284/4257bd58-1c1f-11e6-8b1e-d743d4780d62.png)

**ems types**
![manageiq containers providers](https://cloud.githubusercontent.com/assets/2181522/15315312/6aba58c8-1c1f-11e6-8ce2-b4c192b35779.png)

yaacov commented:
@miq-bot add_label providers/containers, ui, wip

@simon3z this is just for conversation and feedback, I do not yet know if this is the right direction to use.

simon3z commented:
@yaacov any screenshot? Thanks.

yaacov commented:
@simon3z screenshots of the wip, not final.

simon3z commented:
@yaacov you should probably rename "Endpoints" to "Hawkular Metrics".

cc @dclarizio @h-kataria for UI

simon3z commented:
cc @blomquisg for multi-endpoint support.

simon3z commented:
@yaacov let's expose the hawkular metrics for all providers.

yaacov commented:
@simon3z now all providers see the hawkular fields
1. is hawkular hostname a required field ?
2. provider is added to the db, _so this PR does not need changes to db_.

todo:
1. fix the tests to use the new provider functions :-(
2. do update/edit, currently only new is done.

simon3z commented:
> 1. is hawkular hostname a required field ?

No. If missing we shouldn't try to collect metrics (actually later we'll improve trying to auto-detect if metrics are there, but for now let's say that we shouldn't try to collect metrics).

yaacov commented:
@zeari please review, we changed `ems_container` in `routes.rb` to match `ems_cloud`.

zeari commented:
> @zeari please review, we changed ems_container in routes.rb to match ems_cloud.

So @AparnaKarve is making a similar, much more extensive change for `ems_infra` in #7808. 
I think the scope of this PR is much smaller so I would just use #7808 for reference here.

zeari commented:
@AparnaKarve Please review

simon3z commented:
@yaacov I think you can take a step further now and try to use this new endpoint in the metrics collection.

yaacov commented:
> I think you can take a step further now and try to use this new endpoint in the metrics collection.

@simon3z I found metric collection only in kubernetes
https://github.com/ManageIQ/manageiq/pull/7950/files#diff-b04574f4c79fae7080793e5f5ca2d5cf

simon3z commented:
> @simon3z I found metric collection only in kubernetes

That's ok. Thanks.

What else is needed to finalize this? Can you remove `WIP` at this point?

yaacov commented:
> What else is needed to finalize this? Can you remove WIP at this point?

@simon3z all is done :-) ready for review.

yaacov commented:
missing:
1 validation of token
2 route to new is broken

AparnaKarve commented:
@yaacov I just noticed in the screenshot that the `Hostname` field appears twice. Can you remove it from the top if you haven't already done so?

AparnaKarve commented on config/routes.rb:
```diff
@@ -989,6 +989,7 @@
         show_list
         tagging_edit
         tag_edit_form_field_changed
+        ems_container_form_fields
```
With the restful routes changes you've made below in routes.rb, you do not need to specify routes for `new`, `edit` and `show` explicitly anymore, so you can get rid of them.
Refer https://github.com/ManageIQ/manageiq/pull/7808/commits/7925dc71b210be2b0b098828af6ae412c0fae12a for a list of all required changes for restful routes

yaacov replied:
@AparnaKarve :+1: Thanks, I'm currently working to fix the links and routes.

p.s
somehow i lost the route to `/new` any idea :-)

yaacov commented:
> the `Hostname` field appears twice.

@AparnaKarve you are right, the second is the hawkular `hostname`, can you suggest a better name for that field ?

abonas commented:
fyi @pilhuhn @lucasponce (for the hawkular part)

blomquisg commented:
> > the Hostname field appears twice.
> @AparnaKarve you are right, the second is the hawkular hostname, can you suggest a better name for that field ?

@yaacov I think the only thing is that we are typically separating those out to separate tabs on the page.  See [screenshots in this PR from Julian] where he added the multi-endpoint support.  The Default and AMQP endpoints are in different tabs to separate out the fields.

yaacov commented:
@blomquisg Thanks, I will use tabs like in https://github.com/ManageIQ/manageiq/pull/7138

yaacov commented:
update: 
1. fix bearer authentication, need rspecs to be sure.
2. fix most to the routes, need to finish all of them.
3. need to re-do the UI as @blomquisg suggested.

cc @zeari @moolitayer @simon3z

yaacov commented:
Update:
Fixed all routes and links ( that i found :-) ), fixed authentication endpoints.
Need:
1. Re-do the UI as blomquisg suggested.
2. Specs ( and fix what the specs discover ... )

cc @simon3z

simon3z commented:
@miq-bot add_label darga/yes

simon3z commented:
@yaacov we have to finish this up, let us know what's blocking you. What help do you need?

AparnaKarve commented:
@yaacov I was trying to review and test this and got this while trying to add a new Container Provider
```
Error caught: [ActionView::MissingTemplate] Missing partial ems_container/_form, application/_form with {:locale=>[:en], :formats=>[:html], :variants=>[], :handlers=>[:raw, :erb, :html, :builder, :ruby, :haml, :jbuilder, :rjs]}.
```

By any chance, did you forget to include a haml in this PR?

yaacov commented:
@AparnaKarve Yes, thanks, sorry it's not workable and very fluid right now ...
If I will not get them to work today, I will push what I got as is, and hope someone can figure up what I intended them to do :-)

AparnaKarve commented:
@yaacov If you could get the PR in the state that it was originally in when you took those screenshots (a basic workable state, that is), I can take a look at how to add the tabs.
Thanks.

yaacov commented:
@AparnaKarve hi, pushed the missing files ... it is still very WIP but the forms render.

AparnaKarve commented:
@yaacov Cool... will take a look again and keep you posted.

yaacov commented:
@AparnaKarve @simon3z updated the screenshot in the PR description.

AparnaKarve commented:
@yaacov I'm working on a new PR (which is based on your current PR) where I'm trying to use the same `endpoints` haml that is currently being used by the Cloud Provider and the Infrastructure Provider. This way the endpoints related UI is contained in 1 single haml.

I'll also be fixing a few links for proper restful routing.

yaacov commented:
> @yaacov I'm working on a new PR (which is based on your current PR) 

@simon3z @AparnaKarve How do you want to continue with this PR ? 

> I'll also be fixing a few links for proper restful routing.

That is great.

> where I'm trying to use the same endpoints haml that is currently being used by the Cloud Provider and the Infrastructure Provider. This way the endpoints related UI is contained in 1 single haml.

I started by using the one shared haml's, but they get very `if ... else if ... else if ...` and each provider ended up rendering different parts in the same file with very little common parts :-( , that is why I seperated the endpoint haml of the containers from the big haml.

I think the good approach will be to clean the big common haml and separate cloud and infra too, to small hamls.

pilhuhn commented:
The `if...else..` should not check for the provider, but rather use the information of a callback in the provider to find out which tabs to show or not. We had a similar discussion in #8282 and on Gitter with @himdel .

yaacov commented:
@pilhuhn ‎Thanks, I did not know it was talked about. I am waiting for @simon3z and @AparnaKarve inputs to know how to continue.

simon3z commented:
> @pilhuhn ‎Thanks, I did not know it was talked about. I am waiting for @simon3z and @AparnaKarve inputs to know how to continue.

@yaacov I don't understand the question: "@simon3z How do you want to continue with this PR ?".

moolitayer commented:
>> @yaacov I'm working on a new PR (which is based on your current PR)

> @simon3z @AparnaKarve How do you want to continue with this PR ?

@yaacov I think @AparnaKarve meant that you can continue normally with your PR as the intended pr is based on yours and should not effect your work

AparnaKarve commented:
@yaacov @moolitayer Yes, the first commit in my PR would be the same commit in this PR.
Although I had plans on removing the newly created `endpoints` haml in this commit and consolidating  the `ems_container` related code in the endpoints haml that we already have.
Other than that, there are a few bugs related to restful routes and even other bugs that I was planning to fix in my PR.

@yaacov I completely understand your `if...else` concern in the current haml. However the shared haml has worked quite well for us in certain instances -- so there are some advantages.
If you'd rather have a separate haml (like you currently have) for Containers, I'm OK with that as well. 
Let me know.

So here are a few options on how to proceed -
1. Hold on to this PR and I can send you a patch that would fix all the bugs that I find
2. Close this PR. Later I will create a PR with the commit in this PR as my first commit
3. Add me as a "collaborator" and we could collectively work on this

yaacov commented:
> So here are a few options on how to proceed
1. Hold on to this PR and I can send you a patch that would fix all the bugs that I find
2. Close this PR. Later I will create a PR with the commit in this PR as my first commit
3. Add me as a "collaborator" and we could collectively work on this

@AparnaKarve I added you as a "collaborator"

yaacov commented:
> If you'd rather have a separate haml (like you currently have) for Containers, I'm OK with that as well.
Let me know.

@AparnaKarve I think it is best if all ems's will work the same way. So if all use the shared file, containers should do the same. I used a separate file because the code I wrote did not look nice in the shared file :-), but if you can make it work, I think it is better.

AparnaKarve commented:
> but if you can make it work, I think it is better.

@yaacov Thanks - that helps me plan my next step 👍

simon3z commented:
@yaacov will this be available to be merged today before end of sprint 40?

yaacov commented:
>  will this be available to be merged today before end of sprint 40?

1) @simon3z I will fix the rspec fail today, and then decide if we can remove the wip.

cc @AparnaKarve

simon3z commented on app/assets/javascripts/controllers/ems_common/ems_common_form_controller.js:
```diff
@@ -262,6 +270,12 @@ ManageIQ.angular.app.controller('emsCommonFormController', ['$http', '$scope', '
+    } else if ($scope.emsCommonModel.emstype === 'atomic' || $scope.emsCommonModel.emstype === 'atomic_enterprise'){
+      $scope.emsCommonModel.default_api_port = "5000";
+    } else if ($scope.emsCommonModel.emstype === 'kubernetes'){
+      $scope.emsCommonModel.default_api_port = "5000";
```
@yaacov how come you changed these from `8443` to `5000`? The default port for the API endpoint is `8443`.

simon3z commented:
@yaacov I think the "calidate" button disappeared. Is it a known issue?

simon3z commented:
@yaacov I think we should display the hawkular hostname and port in the provider summary as well.
(At the moment the hawkular details are not displayed anywhere).

yaacov commented:
@simon3z 
good news:
1. fixed the port from 5000 to 8443
2. fixed the rspec

not so good:
1. validate is not working, need checking ...
3. display hawkular endpoint, (after validate ?)

simon3z commented on app/assets/javascripts/controllers/ems_common/ems_common_form_controller.js:
```diff
@@ -65,6 +67,7 @@ ManageIQ.angular.app.controller('emsCommonFormController', ['$http', '$scope', '
         $scope.emsCommonModel.openstack_infra_providers_exist = data.openstack_infra_providers_exist;
         $scope.emsCommonModel.default_api_port                = 5000;
         $scope.emsCommonModel.amqp_api_port                   = 5672;
+        $scope.emsCommonModel.hawkular_api_port               = 8443;
```
@yaacov `443` was correct for this one.

yaacov replied:
@simon3z sorry :-) fixed.

simon3z commented:
@yaacov when editing an existing provider, the API port of hawkular in the UI is wrong (in the db it is correct).

juliancheal commented:
not so good:
1. validate is not working, need checking ...
3. display hawkular endpoint, (after validate ?)

If someone could DM me on Gittr or email me some credentials I can look at the validate.

AparnaKarve commented:
@yaacov Do you want me to take a look at Validate while I fix other things?

simon3z commented:
@yaacov how are you handling existing providers? E.g. let's say I created a provider in 4.0, are you making sure that they're supported?

yaacov commented:
> Do you want me to take a look at Validate while I fix other things?

@AparnaKarve very much !!

yaacov commented:
@AparnaKarve do you want me to stop squashing commits from now so you can follow the changes ?

yaacov commented:
> how are you handling existing providers? E.g. let's say I created a provider in 4.0, are you making sure that they're supported?

@simon3z in the rspec i do:
```
expect(@ems.authentication_token("bearer")).to eq('valid-token')
expect(@ems.hostname).to eq('10.10.10.11')
```
and in capture_context:
```
host => hawkular_endpoint ? hawkular_endpoint.hostname : @ext_management_system.hostname,
port => hawkular_endpoint ? hawkular_endpoint.port : worker_class.worker_settings[:metrics_port],
```
but i did not test it or checked it in any other way.

yaacov commented:
> If someone could DM me on Gittr or email me some credentials I can look at the validate.

@juliancheal the problem is not with the validation itself, it's with routing and endpoints ... you do not need an actual provider to see the problem

AparnaKarve commented:
@yaacov Yeah, if you expect to make more changes, keep the commits separate.

yaacov commented:
@simon3z display the hawkular hostname and port in the provider summary

![manageiq containers providers](https://cloud.githubusercontent.com/assets/2181522/15120119/e45ea836-161c-11e6-9d2e-c22854f80d0f.png)

AparnaKarve commented on app/assets/javascripts/controllers/ems_common/ems_common_form_controller.js:
```diff
@@ -65,6 +67,7 @@ ManageIQ.angular.app.controller('emsCommonFormController', ['$http', '$scope', '
         $scope.emsCommonModel.openstack_infra_providers_exist = data.openstack_infra_providers_exist;
         $scope.emsCommonModel.default_api_port                = 5000;
         $scope.emsCommonModel.amqp_api_port                   = 5672;
+        $scope.emsCommonModel.hawkular_api_port               = 443;
```
@yaacov Should the default hawkular port be 443 or 8443? 
A few lines down it is set o 8443.

yaacov replied:
@AparnaKarve see https://github.com/ManageIQ/manageiq/pull/7950#discussion_r62504378

simon3z commented:
Nice improvements @yaacov and @AparnaKarve ! I think we should add validation also for the Hawkular endpoint (when you add a provider).

yaacov commented on app/controllers/mixins/ems_common_angular.rb:
```diff
@@ -392,6 +424,13 @@ def build_credentials(ems, mode)
+        creds[:bearer] = {:auth_key => params[:bearer_password]}
+      end
+
+      ems.update_authentication(creds, :save => (mode != :validate))
```
`ems.update_authentication(creds, :save => (mode != :validate))`

@AparnaKarve  without this line the .`authentication_token("bearer")` is not set :-(
1. is their a better way to do this ?
2. do we need it ?
3. do we need it only for container-manager ?

yaacov commented:
@simon3z @AparnaKarve can we remove the wip ?

zeari commented on app/assets/javascripts/controllers/container_dashboard/container_dashboard_controller.js:
```diff
@@ -61,7 +61,7 @@ miqHttpInject(angular.module('containerDashboard', ['ui.bootstrap', 'patternfly'
         }
         else {
-          id = '/'+ (/ems_container\/show\/(\d+)/.exec($location.absUrl())[1]);
+          id = '/'+ (/\/(\d+)/.exec($location.absUrl())[1]);
```
Why is this part of this patch?

yaacov replied:
@zeari we changed the routes. 
the route was: `/ems_container/show/<:id>`
the route now is: `/ems_container/<:id>`
since correct roude can have only once a `"/<number>" `sequence the new regex will work for the new routes.

simon3z commented:
> @simon3z @AparnaKarve can we remove the wip ?

@yaacov sure. Have you added the support to validate the hawkular endpoint?

yaacov commented:
> Have you added the support to validate the hawkular endpoint?

@simon3z I do not know how / what to validate 
1. that we can get metrics from that hosname and port ?

AparnaKarve commented:
@yaacov Other than @simon3z 's request above (hawkular endpoint validation), there are still 2 spec failures that need to be addressed.
Just wanted to check if those are the only 2 things that are currently unaddressed.

yaacov commented:
@AparnaKarve 2 spec failures ?

AparnaKarve commented:
@yaacov The 2 spec failures related to `EmsInfraController#build_credentials`

yaacov commented:
@AparnaKarve I think the last commit fixed it

yaacov commented:
@AparnaKarve 
I do not know why _bearer_ require the `ems.update_authentication` here, maybe this fix is wrong the the `update_authentication` is missing somewhere else (I did not find such a place) ?

About the hawkular validation. I talked with @simon3z and the way we came up to test _hawkular validity_ is a json with `{"MetricsService":"STARTED"}` answer to `https://[hostname:port]/hawkular/metrics/status`

AparnaKarve commented:
@yaacov The last commit does not seem right.
Looking into the spec failures/`ems.update_authentication` now...

simon3z commented:
@yaacov you need to:

- [ ] rebase
- [ ] fix the tests
- [ ] fix the warning reported by rubocop
- [ ] squash all the patches

AparnaKarve commented:
@yaacov I conducted some tests, and it *does* look like `bearer` needs the `ems.update_authentication` call. So your changes in 11b94d7da032a4e57ca0b3c2a45674c1cc77b003 are correct.

AparnaKarve commented:
@simon3z As far as the rubocop warnings in the hamls go, the ones related to spacing should not be addressed because the spaces around the hash rockets are intentional and promote readability.

Also, do we need to squash the commits? I think separate commits make for better understanding of the code changes.

yaacov commented:
@simon3z 
- [x] rebase - done
- [x] fix the tests - done
- [ ] squash all the patches, @AparnaKarve ?
- [ ] fix the warning reported by rubocop, some can be fixed without making the code unreadable, will fix those.
- [ ] hawkular validation (new PR?)

AparnaKarve commented:
>  squash all the patches, @AparnaKarve ?
@yaacov I wouldn't. 😄 ...but you make the call.

himdel commented:
I'll second **not** squashing .. the commits do seem to convey the intent quite nicely.

Pretty much the only reason to squash I see would be if you had *lots* of commits just rewriting the same place and ..well, experimenting :).

simon3z commented on app/views/ems_container/_main.html.haml:
```diff
@@ -3,6 +3,8 @@
     = render :partial => "shared/summary/textual", :locals => {:title => _("Properties"),
                                                                :items => textual_group_properties}
+    = render :partial => "shared/summary/textual", :locals => {:title => _("Endpoints"),
+                                                               :items => textual_endpoints}                                                           
```
@yaacov please remove the trailing whitespaces in the line above.

simon3z commented:
> Pretty much the only reason to squash I see would be if you had lots of commits just rewriting the same place and ..well, experimenting :).

That's exactly the point, the patches should be pristine to the best of our knowledge, not touch back and forth the same code over and over again. That is good for WIP but not for code that gets reviewed and merged.

Why something that breaks rubocop should be merged, only to then have a PR to "fix fixable rubocop problem". Just fix the relevant PR.

Also remember that you should be able to bisect the branch and always have a working codebase (which is not the case at the moment).

If you want to maintain authorship squash blocks of different authors.

himdel commented:
@simon3z OK, maybe we're talking about different thing when we talk about squashing...

So, my only concern is that a PR with 1 commit touching 20 files is **much** harder to read, than a PR that touches 20 files but over a bunch of thematically separated commits - because the commit messages are important too. And frankly I haven't really looked at the individual commits in detail, but from the commit names ("add missing button", "fix restful routes", "remove duplicate check") it seems perfectly fine to keep it as is.

And please note that I'm definitely not advocating fixing rubocop in a different PR. But I do think it's clearer if it's in a different commit. (And rubocop only warns about the end result, not the intermediate stages..)

That said.. consider this only my 2c please, I don't think it's my call to make :). I'm just saying how we do things :).

simon3z commented:
> So, my only concern is that a PR with 1 commit touching 20 files is much harder to read, than a PR that touches 20 files but over a bunch of thematically separated commits

I agree on that but currently the PR are not thematic. They're just a set of issues found along the way (issues that shouldn't have been introduced in the previous PRs).


> And frankly I haven't really looked at the individual commits in detail, but from the commit names ("add missing button", "fix restful routes", "remove duplicate check") it seems perfectly fine to keep it as is.

@himdel actually I see:

- `haml adjustments and fix for Validate`

What adjustments? Why introducing something broken (Validate) just to fix it later? (I know, different authors)

- `Fix Volumes path`

Why keep it broken in a previous patch?

- `remove duplicate check for ems_container in view_to_url`

Why did we add a duplicate check to begin with?

- `remove unused _endpoints.html.haml, we use the shared file`

This just removes the unused `_endpoints.html.haml`, how's that thematic? Where is the change that is introducing the use of the shared file?

- `default to oldstyle if endpoint-default is missing`

This seems a pretty serious bug that shouldn't have been introduced earlier.

- `use bearer_password and not bearer_token in the spec`

Why fix the spec later? It should have been attached to the relevant change.

- `fix failing tests for ems_common_controller_spec and ems_infra_contro...`

Again, why separate from the relevant code change?

- `fix fixable rubocop problem`

Why not writing rubocop-compliant code to begin with?

I am all for maintaining authorship but this doesn't look professional committed to the official master branch, this is a WIP developer branch.

himdel commented:
@simon3z well, you know, I'd agree on most of the points, but if you think it through, the end result would be that 1 huge commit. And, well. I'd prefer *clarity over form* every time - a PR is not just the end product, it tells a story.. and everybody who would like to do something similar, or who is wondering (say, in a year) why that change happened, they need to see that story, not just a black box..

simon3z commented:
@himdel are you really trying to convince me that reviewing separately 14 patches without any specific theme:

- **containers multi endpoint**
  19 files changed, 339 insertions(+), 74 deletions(-)
- **add hawkular endpoint to the provider summary**
  2 files changed, 10 insertions(+)
- **haml adjustments and fix for Validate**
  6 files changed, 84 insertions(+), 28 deletions(-)
- **Fix Volumes path**
  1 file changed, 1 insertion(+), 1 deletion(-)
- **Refactor Timelines button based on other ems controllers**
  3 files changed, 18 insertions(+), 10 deletions(-)
- **Fix routes to restful in accordions**
  2 files changed, 3 insertions(+), 4 deletions(-)
- **remove duplicate check for ems_container in view_to_url**
  1 file changed, 3 deletions(-)
- **remove unused _endpoints.html.haml, we use the shared file**
  1 file changed, 57 deletions(-)
- **add missing performance button**
  3 files changed, 17 insertions(+), 1 deletion(-)
- **default to oldstyle if endpoint-default is missing**
  1 file changed, 15 insertions(+), 7 deletions(-)
- **use bearer_password and not bearer_token in the spec**
  1 file changed, 1 insertion(+), 1 deletion(-)
- **fix failing tests for ems_common_controller_spec and ems_infra_controller_spec**
  1 file changed, 1 insertion(+), 2 deletions(-)
- **check for ems type before trying to validate**
  1 file changed, 1 insertion(+)
- **fix fixable rubocop problem**
  5 files changed, 22 insertions(+), 18 deletions(-)

for a total of **512** insertions and **206** deletions, constantly moving back and forth to check if a bug that you spotted will be fixed in a subsequent patch on not ...is easier than reviewing the squashed result of **395** insertions and **89** deletions?

Reviewing separately you waste at least **30%** of your time reading lines that are useless as they'll be changed in a later patch.

If you're able to consolidate the PRs in a series of PR that make sense, please do so. Current state doesn't reflect the quality of work that we should be used to.

himdel commented:
@simon3z Well, there is always the option of looking at the whole PR, not individual commits, and only looking at those if you're wondering why something happened :)

But.. I'm not really trying to convince you of anything, and I think I've said all that was needed, now it's up to you guys/gals :).

simon3z commented:
Since this is still in WIP I am re-assigning it to @yaacov 

@miq-bot assign yaacov

yaacov commented:
@simon3z this is an update on done/todo list for this PR
￼ 
Done:
- [x] rebase
- [x] fix the tests
- [x] fix the warning reported by rubocop - fix trailing white space in main.html

Todo:

- [ ] hawkular validation
- [ ] Remove WIP, and squash all the patches by author.

simon3z commented:
> Todo:
> 
>  - [ ] hawkular validation
>  - [ ] Remove WIP, and squash all the patches by author.

@yaacov is it feasible to finish this today?

yaacov commented:
@simon3z 
> is it feasible to finish this today?

I think it is, I am learning how authentication works ... once I understand that, the button will work.

mikedahlgren commented:
I apologize if this is not the correct place to say this, but this feature is a very high priority for the customers I work with and hope that this can be taken care of as soon as possible.

yaacov commented:
@mikedahlgren @simon3z  All the todo's are done.
1. Will remove the wip label after squashing, rechecking, and travis OK.
2. This is a big PR it needs review

yaacov commented:
@simon3z this is an update on done/todo list for this PR

Done:
- [x] rebase
- [x] fix the tests
- [x] fix the warning reported by rubocop - fix trailing white space in main.html
- [x] hawkular validation
- [x] squash all the patches by author.

will remove wip after travis ok

yaacov commented:
@miq-bot remove_label wip

simon3z commented:
@yaacov have you tried the hawkular validation? I get an error:

![screenshot from 2016-05-17 11-16-24](https://cloud.githubusercontent.com/assets/130393/15317281/11639ce2-1c21-11e6-8b45-a61d8e9bb857.png)

```
[----] W, [2016-05-17T11:18:44.436819 #27203:13ef3b8]  WARN -- : MIQ(ManageIQ::Providers::Openshift::ContainerManager#authentication_check_no_validation) type: ["hawkular"] for [1] [Neutrino] Validation failed: incomplete, Missing credentials
```

FYI I am testing this with an existing provider: I had a provider (defined before this PR) and I am editing it.

yaacov commented:
I've tried it on molecule and my private provider, I am checking it now.

![manageiq containers providers](https://cloud.githubusercontent.com/assets/2181522/15319281/c726a720-1c32-11e6-8869-7629e70d3f6c.png)

simon3z commented:
> @simon3z I've tried it on molecule and my private provider, I am checking it now.

@yaacov was the provider created **before** your PR? As I explained I got that error trying to editing an **already existing** provider (defined before your PR).

yaacov commented:
@simon3z 
found a logic bug in `required_credential_fields`, it looked for `userid` that does not have to exist for hawkular. 
i think that is the bug you see. on my side i did not see it because the `userid` field is always set. may be on old versions of manageiq this field was not set to default value for all providers.

yaacov commented:
@enoodle this PR use hawkular client, https://github.com/ManageIQ/manageiq/pull/8472

simon3z commented:
Overall this LGTM now :+1: , @yaacov has addressed all my previous concerns/comments.

@blomquisg can you review the multi-endpoint part for correctness?
@dclarizio @h-kataria is there anyone from your side that can validate the UI side?

Thanks.

AparnaKarve commented:
@dclarizio @simon3z Retested the changes since I last worked on the PR.
Everything looks good. 
OK to merge.

simon3z commented:
OK, we're just missing @blomquisg 

@miq-bot assign chessbyte
